distributed_training/averaging/avg_handler.py

```diff
@@ -0,0 +1,329 @@
+    ) -> Tuple[bool, Dict[str, Any]]:
+        """Process allreduce specifically for validator."""
+        # TODO Weight/gradient validation
+        gradient_averaging_step = None
+        query_tasks = []
+
+        try:
+            # Clip gradients
+            torch.nn.utils.clip_grad_norm_(self.model.parameters(), 1.0)
+
+            # Used for load balancing and scoring
+            self.grad_averager.bandwidth = bandwidth[
+                "download"
+            ]  # TODO Either use average bandwidth or set each time here
+
+            bt.logging.info("Starting Pseudo Gradient Averaging..")
+            gradient_averaging_step = self.grad_averager.step(
+                wait=False,
+                timeout=timeout,
+            )
+
+            async with asyncio.timeout(timeout):
+                # Send AllReduce query to pause miner training and perform global sync
+                query_tasks.append(
+                    dendrite_pool.async_forward(
+                        miner_uids,
+                        [AllReduce() for _ in miner_uids],
+                        timeout=timeout,
+                    )
+                )
+                bt.logging.info(
+                    ":wait: AllReduce Query Sent Out. Waiting for AllReduce to finish.."
+                )
+
+                await asyncio.gather(*query_tasks)
+                bt.logging.info("AllReduce Query Responses Received..")
+
+            self.grad_averager.notify_used_averaged_gradients()
+            bt.logging.success("Finished Averaging Pseudo Gradients!")
+
+            if gradient_averaging_step.done():
+                (
+                    failed_peers,
+                    participating_peers,
+                    modes,
+                    bandwidths,
+                ) = gradient_averaging_step.result()
+
+                initial_weights = self._get_weights_sample()
+                bt.logging.debug(f"Initial weights sample: {initial_weights}")
+
+                # Perform offloaded outer optimization steps
+                bt.logging.info("Performing Outer Optimizer Step..")
+                self.state_averager.step(
+                    increment_epoch=True, optimizer_step=True, zero_grad=False
+                )
+                self.state_averager.update_main_param_after_outer_step()
+                self.state_averager.optimizer.zero_grad()
+                bt.logging.success(
+                    ":white_heavy_check_mark: Finished Outer Optimizer Step!"
+                )
+
+                # Validate weight updates
+                await self._validate_weight_update(initial_weights)
+
+                return {
+                    "failed_peers": failed_peers,
+                    "participating_peers": participating_peers,
+                    "modes": modes,
+                    "bandwidths": bandwidths,
+                }
+
+            else:  # TODO Ensure failing avaerging is handled
```

Comment:
ModelStateError is not a class so the code fails on this line. If averaging fails we should follow this logic:https://github.com/KMFODA/DistributedTraining/blob/2da7e890579043a64df08ddac74cf104a5d7b8f5/distributed_training/validator/forward.py#L270-L363